### PR TITLE
add collectGarbage method for SliverChildDelegate to track which children can be garbage collected

### PR DESCRIPTION
**Pull request by zmtzawqlp**
## Description

flutter has [ResizeImage] to finer control of the size of the image in [ImageCache] and is
generally used to reduce the memory footprint of [ImageCache], but it's not enough. In fact, at the most time, we can't get image size(width/height) before image is aleady, so we can't set cache width and cache height in the moment.

it should provide an opportunity to clear image cache when children can be garbage collected. user  will get indexes from [collectGarbage] method and clear image cache by using [ImageProvider.evict()].

## Related Issues

https://github.com/flutter/flutter/issues/19558

## Tests

I added the following tests:

* ‘collectGarbage has correct indices’

* prepare ListView(itemExtent: 100.0,cacheExtent: 0.0,) with TestSliverChildListDelegate which will add log into gcLog when collectGarbage is called.
* drag ListView -600.0 and expected log will be 'collectGarbage indexes=[0, 1, 2, 3, 4, 5]'
* drag ListView 100.0 and expected log will be 'collectGarbage indexes=[11]'

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

### CODE REVIEW

**Comment by zmtzawqlp** [comment]
@googlebot I signed it!

**Comment by goderbauer** [comment]
The issue #19558 has been closed with a different fix. Is this change still necessary?

/cc @liyuqian

**Comment by zmtzawqlp** [comment]
@liyuqian @goderbauer 

without collect garbage, the memory increase fast , reach 1.5b at android and crash at ios.
memory1
memory2

when i evict image cache by this method, memory is stable.
memory3
memory4

in our actual development, we can't improve memory with ResizeImage due to we also care about the quality of image, reduce cache image width and height will reduce the quality . on the other hand, not everyone can easy to get the size of the image before it's decoded, we can't define cache image width/height without aspect ratio of image.

i have added this method in my extendef flutter packages [extended_list](https://pub.flutter-io.cn/packages/extended_list) ,but  i think it should provide an official solution to improve big long image list.

following are full demo of this issue. in this demo ,i use both collectGarbage and ResizeImage 
https://github.com/fluttercandies/extended_image/blob/master/example/lib/pages/photo_view_demo.dart
 
memory5

I'm look forward that you can take a look when you have a time. thanks.

**Comment by liyuqian** [comment]
> The issue #19558 has been closed with a different fix. Is this change still necessary?
> 
> /cc @liyuqian

@goderbauer : yes, this, or something similar to this, is still needed to reduce the peak memory usage since https://github.com/flutter/flutter/issues/19558 is only about decreasing the memory usage in the end of the image list scrolling, but not during the scrolling.

One potential improvement for this PR is to not leave `collectGarbage` as a responsibility for the user to override and implement. The Flutter framework should automatically call `evict` when there's a memory pressure.

My understanding is that Flutter is already calling `_checkCacheSize` whenever a new image is inserted to remove cache if the max size is reached. However, `_checkCacheSize` doesn't remove pending images while `evict` does. I think that's why this PR can still reduce the memory significantly.

One way to automatically clear pending images could be: (1) whenever an image is moved offscreen, add its key to something like `ImageCache.unusedImageKeys`; (2) whenever a memory pressure is detected (i.e. max cache size is reached), remove all pending images in `ImageCache.unusedImageKeys`.

Also, it would be nice for this PR to write a memory test to verify the improvement, and guard future regressions. https://github.com/flutter/flutter/pull/46033 is an example memory test. Right now, we've only detected memory usages at a few time points. @terrylucas is working on a new tool to persistent the whole timeline of the memory usage during the app's lifespan https://github.com/flutter/flutter/issues/44013. I think that new tool could be helpful for this test.

Also CC @dnfield @chinmaygarde @jonahwilliams as they've worked on image cache memory issues before.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 1]
```diff
@@ -367,6 +371,7 @@ abstract class RenderSliverMultiBoxAdaptor extends RenderSliver
       childParentData._keptAlive = true;
     } else {
       assert(child.parent == this);
+      indexes.add(childParentData.index);
```
Comment (dnfield) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: I'm confused here why `removeChild` below doesn't do what we want here.  Calling `collectGarbage` later is just going to invoke `removeChild` again right?

Comment (dnfield) [reply packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: Ah, I see now from your other example.

**Comment by dnfield** [comment]
Is there some way we could update the elements used for images so that when `detachRenderObject` is called on them, we optionally evict the image from the cache?

**Comment by dnfield** [comment]
The idea being that when the current garbage collection routines run, `detachRenderObject` gets called on the element child.

**Comment by zmtzawqlp** [comment]
@liyuqian @dnfield 
automatically clear pending images when detachRenderObject  looks good. but we can't know whether we should clear cache for user's case. 
clear image cache should only if your image has local disc cache. if not, image will download again when scroll back.
as i see, auto clear cache is not safe. for example, click image in list and pop new page to view image( they use the same provider). then push page back , view image page will dispose but we can't clear image cache at that moment.

@liyuqian  
i don't konw how to make a memory test likes https://github.com/flutter/flutter/pull/46033 for this pr,  it likes the different case. can you help me if you have time? thanks.

**Comment by dnfield** [comment]
I'm imagining we would have a new property on Image to say whether it should be cleared on detach

**Comment by dnfield** [comment]
In this case, I actually think we should not even put these images in the cache.  The render object will hold the image in memory until it's off screen.  

If the concern is about network images, we should probably just have a better way to cache what's important about network images (where we may not be able to/want to cache the entire texture, but we do want to cache what we got from the network, perhaps to disk).

**Review by dnfield — CHANGES_REQUESTED** [review]
I'm not sure this is the right API. I think it will be very difficult to use correctly and will be unlikely to help developers do the right thing.  In particular, it seems to encourage caching images you don't actually want to cache.

**Comment by zmtzawqlp** [comment]
> I'm not sure this is the right API. I think it will be very difficult to use correctly and will be unlikely to help developers do the right thing. In particular, it seems to encourage caching images you don't actually want to cache.

yes, you are right. It would be better if we could resolve it from origin of this issue. In fact, this method is more like didFinishLayout method. they told user which children can be garbage collected and which are ready. i will keep an eye this, thanks

**Comment by liyuqian** [comment]
@zmtzawqlp : sure, I'm very happy to pair with you to see how to write a proper memory test. It's a less explored area inside Flutter so I'll probably learn with you. For now, I'm still waiting for the fixing of https://github.com/flutter/flutter/issues/44013 . I think it will make this test a little easier.

**Comment by dnfield** [comment]
And in case it's not clear from my comments I do think the approach is interesting! IT definitely highlights some problems we have and provided some good discussion of how to fix it!  Thanks for it!

**Comment by dnfield** [comment]
At this point I think all the relevant issues have been filed - let's follow up on those and close this for now. If we discover a new need for API like this we can reopen.

**Pull request closed without merging**